packages/@ember/-internals/glimmer/lib/renderer.ts

```diff
@@ -378,108 +724,112 @@ export class Renderer {
-      this._context,
-      this._owner,
+      this.state.runtime,
+      this.state.compilation,
+      this.state.owner,
       this._rootTemplate,
       self,
       target,
       dynamicScope,
-      this._builder
-    );
-    this._renderRoot(rootState);
-  }
-
-  rerender(): void {
-    this._scheduleRevalidate();
-  }
-
-  register(view: any): void {
-    let id = getViewId(view);
-    assert(
-      'Attempted to register a view with an id already in use: ' + id,
-      !this._viewRegistry[id]
+      this.state.builder
     );
-    this._viewRegistry[id] = view;
+    this.state.renderRoot(rootState, this);
   }
 
-  unregister(view: any): void {
-    delete this._viewRegistry[getViewId(view)];
-  }
-
-  remove(view: Component): void {
-    view._transitionTo('destroying');
-
-    this.cleanupRootFor(view);
-
-    if (this._isInteractive) {
-      view.trigger('didDestroyElement');
-    }
-  }
-
-  cleanupRootFor(view: unknown): void {
+  cleanupRootFor(component: ClassicComponent): void {
     // no need to cleanup roots if we have already been destroyed
-    if (this._destroyed) {
+    if (isDestroyed(this)) {
       return;
     }
 
-    let roots = this._roots;
+    let roots = this.state.roots;
 
     // traverse in reverse so we can remove items
     // without mucking up the index
-    let i = this._roots.length;
+    let i = roots.length;
     while (i--) {
       let root = roots[i];
       assert('has root', root);
-      if (root.isFor(view)) {
+      if (root.type === 'classic' && root.isFor(component)) {
         root.destroy();
         roots.splice(i, 1);
       }
     }
   }
 
-  destroy() {
-    if (this._destroyed) {
-      return;
+  remove(view: ClassicComponent): void {
+    view._transitionTo('destroying');
+
+    this.cleanupRootFor(view);
+
+    if (this.state.isInteractive) {
+      view.trigger('didDestroyElement');
     }
-    this._destroyed = true;
-    this._clearAllRoots();
   }
 
-  getElement(view: View): Nullable<Element> {
+  get _roots() {
+    return this.state.debug.roots;
+  }
+
+  get _inRenderTransaction() {
+    return this.state.debug.inRenderTransaction;
+  }
+
+  get _isInteractive() {
+    return this.state.debug.isInteractive;
+  }
+
+  get _context() {
+    return this.state.compilation;
+  }
+
+  register(view: any): void {
+    let id = getViewId(view);
+    assert(
+      'Attempted to register a view with an id already in use: ' + id,
+      !this._viewRegistry[id]
+    );
+    this._viewRegistry[id] = view;
+  }
+
+  unregister(view: any): void {
+    delete this._viewRegistry[getViewId(view)];
+  }
+
+  getElement(component: View): Nullable<Element> {
     if (this._isInteractive) {
-      return getViewElement(view);
+      return getViewElement(component);
     } else {
       throw new Error(
         'Accessing `this.element` is not allowed in non-interactive environments (such as FastBoot).'
       );
     }
   }
 
-  getBounds(view: View): {
+  getBounds(component: View): {
```

Comment:
what's a "view"? 😅 🙈

Reply:
TL;DR almost every use of the word "view" means "classic component". I considered renaming all `view` to `classicComponent` but it was expanding the scope of this PR a lot. That said, it's actually really clarifying when you see an internal API that literally only takes classic components. It means it must not be all that fundamental to the rendering process since empirically we actually do render a lot of Glimmer components.

That said, it also illustrates a way in which route templates are special: they rely on mounting via APIs on the renderer, and (before this PR) there was literally no way to mount a root component or an outlet component in Ember that was implemented as a Glimmer component.